teps/0117-manage-pipelinerun-concurrency.md

```diff
@@ -0,0 +1,368 @@
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+Manage pipeline run concurrency via configurable strategies.
+
+## Motivation
+There are situations where pipelines can overlap, and they should not. For instance, a pr is created and that triggers ci to so a PipelineRun is created, a developer notices a small typo, pushed again, and a new PipelineRun is triggered, the first PipelineRun is no longer relevant, but potentially still running.  
+Another example is managing available shared resources, lets say tekton triggers an external device and there are a set available amount, or simply resource constrained clusters. 
+
+
+2 proposed cancel strategies. 
+- ZeroOverlap:
+  - Pipelineruns with the same key do not overlap, so if one is running and a new one is scheduled, the former is canceled.
+  - Configure cancel status ( with finally, stop, cancel )
+- Queues:
+  - Queues have a max run ( how many can run at the same time )
+  - fifo
+  - Queue max can be over-ridden, but is always the last pipelinerun created.     
+
+
+Strategies can be mixed together BUT their concurrency keys cannot be the same ( cancel eachother out ).
+
+Mixing strategies would allow for better resource management for teams. For instance, 3 teams are working on a mobile application with only 5 test devices so our max queue size is 5. If team 1 works in really fast bursts, 
+they might clog the queue with small pr changes. Leaving team 2 & 3 only able to get a build in when lucky. With zeroOverlap, and a queue, a single pr with many quick changes will only ever occupy one slot in the queue.    
+
+### Requirements
+- PipelineRuns are canceled based on the strategy defined.
+- Simple interface set at the PipelineRun manifest.
+- New PipelineRun should wait for the canceled one to finish when applicable.
+- All PipelineRuns follow fifo patterns. 
+
+## Proposal
+Adding a key on the PipelineRun manifest seems like the most appropriate, and simple solution.  
```

Comment:
I'm wondering if it could make sense to have any of this defined on the Pipeline instead. for example, in the case where you want only one CI run at a time:

```
kind: Pipeline
spec:
  concurrency:
    zeroOverlap:
      key: $(params.pull-request-id)
```

Reply:
ah this is interesting, like having the idea of overlap closer to the actual thing you don't want overlapped ( for lack of better phrasing lolol 🙈 )? There is something that feels nice and clean about that, but I wonder if in practice it will be hard for users to remember where one config goes vs another?

Reply:
Yeah! The reason I suggest this is because the Pipeline author knows whether concurrency is appropriate for the pipeline, and this will allow users to not have to set concurrency controls on every pipelinerun for that pipeline. I'm sure we could explore adding this to the PipelineRun later if there's a clear need. What do you mean by it being hard to remember where one config goes vs another?

Reply:
All I meant on the topic of remembering where configs might go, if the zeroOverlap is on the pipeline, but the queue config is on the PipelineRun, then there are 2 areas that are managing a similar topic ( concurrency ). That being said, I dont think it is a hard stop or anything, just a thought.

Reply:
I think we could probably add both options to the Pipeline spec for now (like in the alternative syntax example I put below) and then we could always add `concurrency` to PipelineRun later if needed. Do you think the "queue" option is something you would want to specify on the PipelineRun?

Reply:
So there is a debate of "is concurrency an authoring or a runtime concern ?" 🙃. 

> The reason I suggest this is because the Pipeline author knows whether concurrency is appropriate for the pipeline, and this will allow users to not have to set concurrency controls on every pipelinerun for that pipeline.

I partly disagree on this. It might be true for some usecase, but I am not sure this would be the majority, at all.
If I have a Go `Pipeline` (build, test, publish an image or something), as an author, I do not need nor want to set concurrency on this, as I don't really have to care about it. **But** as a consumer of this Pipeline, I may want to have only 1 PipelineRun for this `Pipeline` to run **per Go project**.
In addition, putting this on the `Pipeline` spec might reduce `Pipeline` "share-ability". Taking that Go `Pipeline` again, and assuming it's available in a catalog (or better, through the hub) ; if I want to achieve what I described above, I actually need to "duplicate" the `Pipeline` or *augment* it using, for example, Kustomize.

My take on this is almost the opposite as @lbernick 👼🏼 🙃. I'd rather start by having this on the `PipelineRun` spec as it's closer to a "runtime" concern than an authoring one. And *then* I would explore the *in `Pipeline`*. Not having it on `PipelineRun` also reduce a bit it's "potential", because then any tool (or user) using embedded specs (like [`pipelines-as-code`](https://pipelinesascode.com/)) will not benefit from it at all.

Reply:
Thinking about this a bit more today-- what would we do if two PipelineRuns had the same concurrency key but different concurrency strategies? I think the concurrency strategy has to be defined somewhere other than the PipelineRun-- either on the Pipeline or in a configmap or new CRD.

Reply:
Aren't we thinking too much about this though ? 
Who are we targeting for this feature ? Do we assume the majority of the usage of this features will be tools (pipeline-as-code, team's tooling, …) or end-users ? Can this be documented as an actual "known limitation" ? Can we make those keys unique maybe (compute and append a hash on the referenced pipeline or embedded pipeline definition, …) ?

Cluster-level configuration is a bit tricky. It *works* for some cases, mainly, for cluster-admins to limit the number of "resources" consumed by tekton, but teams on their own namespaces will probably need to handle those per namespace or "project" or pipeline.

Reply:
To detail this a bit more:
- on the "making things unique", the controller could take the user's provided concurrency-key, but when "computing" it, also "hashing" the pipelinespec the pipelinerun refers to (be it through pipelineref or embedded spec). That way, it wouldn't "conflict" if 2 PR refers to different pipeline (or an update on a pipeline) but they same concurrency key. This comes with its own set of challenges/problems but we could keep that approach in mind while experimenting.
- the "main" assumption I am making here though, is that in most of the cases, the users doesn't create the PipelineRun himself (aka authoring it), tools do it for them (clients like tkn, tools like pipelines-as-code, home-grown tool, ibm stuff, …). And in the small amount of cases where the user author them, it's probably fine to have this known limitation/issue.

In a gist, I really do think, while experimenting, while it's in alpha & all, we can not overthink this, and see with "usage" if it really is a problem or not.